### PR TITLE
Automation of Online Reads Balancer via Balancer module

### PR DESCRIPTION
**Pull request by pdhiran**
1. Bug automation : https://bugzilla.redhat.com/show_bug.cgi?id=2237574
2. Feature automation of Reads balancer via Balancer module
3. Few -ve scenarios added.

Pass log for Bugzilla and -ve scenario addition : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-KIT9ZE

### CODE REVIEW

File: tests/rados/test_online_reads_balancer.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_obj = RadosOrchestrator(node=cephadm)
+    pool_obj = PoolFunctions(node=cephadm)
+
+    regex = r"\s*(\d.\d)-rhel-\d"
+    build = (re.search(regex, config.get("build", config.get("rhbuild")))).groups()[0]
+    if not float(build) >= 8.0:
+        log.info(
+            "Test running on version less than 7.0, skipping verifying Reads Balancer functionality"
```
Comment (pdhiran) [inline tests/rados/test_online_reads_balancer.py:33]: Update comment

File: tests/rados/test_online_reads_balancer.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                log.error(
+                    f"Read score on pool {pool} has increased post enabling Balancer based reads balancing"
+                )
+                raise Exception(
+                    "Balancer score higher upon enabling reads balancing. Fail"
+                )
+        log.debug("Completed verifying reads balancer scores on all the pools")
+
+        test_pool = "balancer_test_pool"
+        if not rados_obj.create_pool(pool_name=test_pool, pg_num=16):
```
Comment (pdhiran) [inline tests/rados/test_online_reads_balancer.py:121]: Write data into pool.

File: tests/rados/test_online_reads_balancer.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            log.info("Verified balancer mode. Could not be enabled. Proceeding")
+
+        log.debug(
+            "Setting config to allow clients to perform read balancing on the clusters."
+        )
+        config_cmd = "ceph osd set-require-min-compat-client reef"
+        rados_obj.run_ceph_command(cmd=config_cmd)
+        time.sleep(10)
+
+        if not rados_obj.enable_balancer(balancer_mode="upmap-read"):
```
Comment (neha-gangadhar) [inline tests/rados/test_online_reads_balancer.py:90]: can we also validate "read" mode alone? along with "upmap-read"? May be in next PR and add backlog tracker.

**Review by neha-gangadhar — COMMENTED** [review]
lgtm

File: ceph/rados/core_workflows.py [hunk 4]
```diff
@@ -725,14 +725,24 @@ def enable_balancer(self, **kwargs) -> bool:
...
         # Turning on the balancer on the system
         cmd = "ceph balancer on"
         self.node.shell([cmd])
 
+        # Setting the mode for the balancer. Available modes: none|crush-compat|upmap|upmap-read|read
+        balancer_mode = kwargs.get("balancer_mode", "upmap")
+        cmd = f"ceph balancer mode {balancer_mode}"
+        self.node.shell([cmd])
+
+        cmd = "ceph balancer status"
```
Comment (harshkumarRH) [inline ceph/rados/core_workflows.py:737]: If the mode is being changed, does there need to be a slight delay before validation?

File: tests/rados/test_online_reads_balancer.py (hunk 1 above)
Comment (harshkumarRH) [reply tests/rados/test_online_reads_balancer.py:33]: Given that the test is being included in test suites Squid and above, do we really need this block of code?

File: ceph/rados/core_workflows.py [hunk 5]
```diff
@@ -751,6 +761,38 @@ def enable_balancer(self, **kwargs) -> bool:
         log.info(f"the balancer status is \n {out}")
         return True
 
+    def get_read_scores_on_cluster(self) -> dict:
```
Comment (harshkumarRH) [inline ceph/rados/core_workflows.py:764]: I think we can enhance this module or write a new one to fetch the read score for a particular pool. Reason being in the test case, there are at least 2 instances where we are fetching the read score for pool `test-pool`, but for doing so we are first gathering the stats for all the pools from the cluster

Comment (pdhiran) [reply ceph/rados/core_workflows.py:764]: Sounds Good. I'll add a backlog task to make changes for the method.

File: tests/rados/test_online_reads_balancer.py (hunk 3 above)
Comment (pdhiran) [reply tests/rados/test_online_reads_balancer.py:90]: Made the changes. Made the balancer mode a configurable param from the suite file.

File: ceph/rados/core_workflows.py (hunk 4 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py:737]: Added delay.

**Comment by pdhiran** [comment]
@red-hat-storage/ceph-qe-rados , made the changes. Please review.

**Review by harshkumarRH — APPROVED** [review]
Looks good!

**Review by neha-gangadhar — APPROVED** [review]
Approved

**Pull request merged**